### ISSUE TITLE
GoogleMap public api - no methods work

### ISSUE THREAD

**Issue by dyyylan** [tomchentw/react-google-maps#133]
I'm having an issue accessing any of the methods that should be exposed on the GoogleMap component.

In any of these methods, `this.state` is undefined: https://github.com/tomchentw/react-google-maps/blob/0e417015e3444e35c1c7cf4c5f325bfeabcdb584/src/GoogleMap.js#L32-L65

**Comment by tomchentw** [comment]
I think it works on the `<GoogleMap>` component in the example. Could you double confirm?

Code: https://github.com/tomchentw/react-google-maps/blob/67895ab6527e59354c7b6a414fc5651adec97483/examples/gh-pages/scripts/components/GeojsonToComponents.js#L74-L101
Demo: https://tomchentw.github.io/react-google-maps/#geojson

**Comment by dyyylan** [comment]
`this.refs.map.getZoom()` doesn't work either for me. Any of the methods off of `this.refs.map` give me an error in the GoogleMap component, `this.state` is undefined.

Here's the code I'm using:
https://gist.github.com/dyyylan/849d073d7ab16e9a3860

I tried cloning the repo to set up the gh-pages and see if it worked locally, but none of the pages will load.

**Comment by tomchentw** [comment]
So you can't load `examples/gh-pages` locally? Weird. What's your `node`, `npm` version?

The gist looks okay though. However, in componentDidUpdate, could you try to call `this.refs.map.fitBounds()` directly? I'm not sure if that's the case.

**Comment by dyyylan** [comment]
Same error. The line highlighted is this here:

```
{
    key: "fitBounds",
    value: function fitBounds(bounds) {
      return this.state.map.fitBounds(bounds);
    }
  }
```

I'm running node 4.1.2 and npm 2.14.4

**Comment by tomchentw** [comment]
Did you load google.mas script before React?

```
/*
 * This is the modify version of:
 * https://developers.google.com/maps/documentation/javascript/examples/event-arguments
 *
 * Add <script src="https://maps.googleapis.com/maps/api/js"></script> to your HTML to provide google.maps reference
 */
```

**Comment by dyyylan** [comment]
Yes. I should mention the rest of the map stuff is working great, it's just these couple exposed methods that don't work.

**Comment by tomchentw** [comment]
It's really weird. I updated the `GeojsonToComponents` module to include `componentDidUpdate` and it works great.

**Comment by dyyylan** [comment]
Ok I did some more testing and I was able to reproduce what is causing the problem.

In your examples, the map is loaded as soon as the component is mounted.

If something prevents the component from mounting immediately (i.e. I load data and show a spinner gif before the map is shown on the page), something in the GoogleMap state is not bound

I created a repo so you can see how this happens:
https://github.com/dyyylan/map-error

Do you have any suggestions to work around this?

**Comment by tomchentw** [comment]
I tried to update the `Map` component to:

```js
    componentDidUpdate() {
        if (this.refs.map) {
            console.log('Zoom level:', this.refs.map.getZoom());
        }
    }
```

But it doesn't work. Still the error of `Uncaught TypeError: Cannot read property 'getZoom' of undefined` due to `this.state.map` is `undefined`.

**Comment by tomchentw** [comment]
@dyyylan I had a workaround here: https://github.com/dyyylan/map-error/compare/master...tomchentw:master

**Comment by PanJ** [comment]
@tomchentw I'm having exact same problem but I cannot view your workaround you posted.

**Comment by sijovijayan** [comment]
@dyyylan @tomchentw , Can you change the value of 'ref' attribute in the component render block ?
Use something like the following

         ref="glmap"

Please let me know the result.

**Comment by dyyylan** [comment]
Sorry, I cleaned up some repositories I wasn't using and deleted my fork of this project. @PanJ, the workaround was to just create a wrapper component for the map. Whichever component the map is in, it must render when the component mounts, or you won't get the api exposed. That was the cause of my issue

**Comment by beeant** [comment]
I'm having this problem too after upgrading to newer version.

`Uncaught TypeError: Cannot read property 'fitBounds' of undefined`

**Comment by tomchentw** [comment]
A reproducible jsfiddle would be appreciated! @beeant

**Comment by beeant** [comment]
@tomchentw sorry, it works after I moved options to `options` props

**Comment by tomchentw** [comment]
Separating `google.maps` instance creation with React component mounting completely solves this problem. See #157

**Comment by tomchentw** [comment]
Released [v4.6.0](https://github.com/tomchentw/react-google-maps/releases/tag/v4.6.0)
